numpy/__init__.pyi

```diff
@@ -1,15 +1,43 @@
-# very simple, just enough to start running tests
-#
 import builtins
-from typing import Any, Mapping, List, Optional, Tuple, Union
+
+from typing import (
+    Any, Dict, Iterable, List, Optional, Mapping, Sized,
+    SupportsInt, SupportsFloat, SupportsComplex, SupportsBytes, SupportsAbs,
+    Tuple, Union,
+)
+
+import sys
 
 from numpy.core._internal import _ctypes
 
 _Shape = Tuple[int, ...]
 
+# Anything that can be coerced into numpy.dtype. To avoid recursive
+# definitions, any nested fields are required to be castable to a dtype object
+# are typed as Any.
+# Reference: https://docs.scipy.org/doc/numpy/reference/arrays.dtypes.html
+_DtypeLike = Union[
+    dtype,
+    type,  # TODO: enumerate np.generic types and Python scalars
+    # TODO: add a protocol for anything with a dtype attribute?
+    None,
+    str,
+    Tuple[Any, int],
+    Tuple[Any, _Shape],
+    List[Union[Tuple[Union[str, Tuple[str, str]], Any],
+               Tuple[Union[str, Tuple[str, str]], Any, _Shape]]],
+    Dict[str, Any],
+    Tuple[Any, Any]]
```

Comment:
I think each of these could do with a comment giving an example of the case they aim to cover